From 6415dcfcc8dd517ac1085756c6c1fc0e226df32f Mon Sep 17 00:00:00 2001
From: "Andrew M. Bradley" <ambrad@cs.stanford.edu>
Date: Fri, 22 Sep 2023 16:33:53 -0600
Subject: [PATCH] dc3d.f: Modify original source in four ways.

1. REAL*4 -> REAL*8.
2. Clean up COMMON blocks for consistent sizing.
3. Add OMP THREADPRIVATE for COMMON blocks.
4. Add SPACE argument, 'f'/'F' or 'h'/'H' for, respectively, full- and
   half-spaces.
---
 extern/dc3d.f | 36 ++++++++++++++++++++++++++++--------
 1 file changed, 28 insertions(+), 8 deletions(-)

diff --git a/extern/dc3d.f b/extern/dc3d.f
index 596e292..cd2f3d8 100644
--- a/extern/dc3d.f
+++ b/extern/dc3d.f
@@ -1,8 +1,10 @@
-      SUBROUTINE  DC3D0(ALPHA,X,Y,Z,DEPTH,DIP,POT1,POT2,POT3,POT4,
+      SUBROUTINE  DC3D0(SPACE,ALPHA,X,Y,Z,DEPTH,DIP,POT1,POT2,POT3,POT4,
      *               UX,UY,UZ,UXX,UYX,UZX,UXY,UYY,UZY,UXZ,UYZ,UZZ,IRET)
       IMPLICIT REAL*8 (A-H,O-Z)
-      REAL*4   ALPHA,X,Y,Z,DEPTH,DIP,POT1,POT2,POT3,POT4,
+      REAL*8   ALPHA,X,Y,Z,DEPTH,DIP,POT1,POT2,POT3,POT4,
      *         UX,UY,UZ,UXX,UYX,UZX,UXY,UYY,UZY,UXZ,UYZ,UZZ
+      CHARACTER SPACE
+      INTEGER*4 IRET
 C
 C********************************************************************
 C*****                                                          *****
@@ -34,9 +36,10 @@ C*****               :   =0....NORMAL
 C*****               :   =1....SINGULAR
 C*****               :   =2....POSITIVE Z WAS GIVEN
 C
-      COMMON /C1/DUMMY(8),R
+      COMMON /C1/DUMMY(8),R,DUMMY1(15)
       DIMENSION  U(12),DUA(12),DUB(12),DUC(12)
       DATA  F0/0.D0/
+!$OMP THREADPRIVATE(/C1/)
 C-----
       IRET=0
       IF(Z.GT.0.) THEN
@@ -79,6 +82,7 @@ C-----
 C=======================================
 C=====  IMAGE-SOURCE CONTRIBUTION  =====
 C=======================================
+      IF (SPACE.EQ.'F' .OR. SPACE.EQ.'f') GO TO 334
       DD=DEPTH-Z
       CALL DCCON1(XX,YY,DD)
       CALL UA0(XX,YY,DD,PP1,PP2,PP3,PP4,DUA)
@@ -90,6 +94,7 @@ C-----
         IF(I.GE.10) DU=DU+DUC(I-9)
         U(I)=U(I)+DU
   333 CONTINUE
+  334 CONTINUE
 C=====
       UX=U(1)
       UY=U(2)
@@ -141,6 +146,7 @@ C
      *           UY,VY,WY,UZ,VZ,WZ
       DATA F0,F1,F3/0.D0,1.D0,3.D0/
       DATA PI2/6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C1/)
 C-----
       DO 111  I=1,12
   111 U(I)=F0
@@ -243,6 +249,7 @@ C
       DATA F0,F1,F2,F3,F4,F5,F8,F9
      *        /0.D0,1.D0,2.D0,3.D0,4.D0,5.D0,8.D0,9.D0/
       DATA PI2/6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C1/)
 C-----
       C=D+Z
       RD=R+D
@@ -361,10 +368,12 @@ C***** OUTPUT
 C*****   U(12) : DISPLACEMENT AND THEIR DERIVATIVES
 C
       COMMON /C0/ALP1,ALP2,ALP3,ALP4,ALP5,SD,CD,SDSD,CDCD,SDCD,S2D,C2D
-      COMMON /C1/P,Q,S,T,XY,X2,Y2,D2,R,R2,R3,R5,QR,QRX,A3,A5,B3,C3
+      COMMON /C1/P,Q,S,T,XY,X2,Y2,D2,R,R2,R3,R5,QR,QRX,A3,A5,B3,C3,
+     *        DUMMY(6)
       DATA F0,F1,F2,F3,F5,F7,F10,F15
      *        /0.D0,1.D0,2.D0,3.D0,5.D0,7.D0,10.D0,15.D0/
       DATA PI2/6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C1/)
 C-----
       C=D+Z
       Q2=Q*Q
@@ -468,12 +477,14 @@ C=========================================
       ENDIF
       RETURN
       END
-      SUBROUTINE  DC3D(ALPHA,X,Y,Z,DEPTH,DIP,
+      SUBROUTINE  DC3D(SPACE,ALPHA,X,Y,Z,DEPTH,DIP,
      *              AL1,AL2,AW1,AW2,DISL1,DISL2,DISL3,
      *              UX,UY,UZ,UXX,UYX,UZX,UXY,UYY,UZY,UXZ,UYZ,UZZ,IRET)
       IMPLICIT REAL*8 (A-H,O-Z)
-      REAL*4   ALPHA,X,Y,Z,DEPTH,DIP,AL1,AL2,AW1,AW2,DISL1,DISL2,DISL3,
+      REAL*8   ALPHA,X,Y,Z,DEPTH,DIP,AL1,AL2,AW1,AW2,DISL1,DISL2,DISL3,
      *         UX,UY,UZ,UXX,UYX,UZX,UXY,UYY,UZY,UXZ,UYZ,UZZ
+      CHARACTER SPACE
+      INTEGER*4 IRET
 C
 C********************************************************************
 C*****                                                          *****
@@ -503,10 +514,11 @@ C*****               :   =0....NORMAL
 C*****               :   =1....SINGULAR
 C*****               :   =2....POSITIVE Z WAS GIVEN
 C
-      COMMON /C0/DUMMY(5),SD,CD
+      COMMON /C0/DUMMY(5),SD,CD,DUMMY1(5)
       DIMENSION  XI(2),ET(2),KXI(2),KET(2)
       DIMENSION  U(12),DU(12),DUA(12),DUB(12),DUC(12)
       DATA  F0,EPS/ 0.D0, 1.D-6 /
+!$OMP THREADPRIVATE(/C0/)
 C-----
       IRET=0
       IF(Z.GT.0.) THEN
@@ -590,6 +602,7 @@ C-----
 C=======================================
 C=====  IMAGE-SOURCE CONTRIBUTION  =====
 C=======================================
+      IF (SPACE.EQ.'F' .OR. SPACE.EQ.'f') GO TO 335
       D=DEPTH-Z
       P=Y*CD+D*SD
       Q=Y*SD-D*CD
@@ -646,6 +659,7 @@ C-----
 C-----
   333 CONTINUE
   334 CONTINUE
+  335 CONTINUE
 C=====
       UX=U(1)
       UY=U(2)
@@ -696,6 +710,7 @@ C
       COMMON /C2/XI2,ET2,Q2,R,R2,R3,R5,Y,D,TT,ALX,ALE,X11,Y11,X32,Y32,
      *           EY,EZ,FY,FZ,GY,GZ,HY,HZ
       DATA F0,F2,PI2/0.D0,2.D0,6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C2/)
 C-----
       DO 111  I=1,12
   111 U(I)=F0
@@ -780,6 +795,7 @@ C
       COMMON /C2/XI2,ET2,Q2,R,R2,R3,R5,Y,D,TT,ALX,ALE,X11,Y11,X32,Y32,
      *           EY,EZ,FY,FZ,GY,GZ,HY,HZ
       DATA  F0,F1,F2,PI2/0.D0,1.D0,2.D0,6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C2/)
 C-----
       RD=R+D
       D11=F1/(R*RD)
@@ -898,6 +914,7 @@ C
       COMMON /C2/XI2,ET2,Q2,R,R2,R3,R5,Y,D,TT,ALX,ALE,X11,Y11,X32,Y32,
      *           EY,EZ,FY,FZ,GY,GZ,HY,HZ
       DATA F0,F1,F2,F3,PI2/0.D0,1.D0,2.D0,3.D0,6.283185307179586D0/
+!$OMP THREADPRIVATE(/C0/, /C2/)
 C-----
       C=D+Z
       X53=(8.D0*R2+9.D0*R*XI+F3*XI2)*X11*X11*X11/R2
@@ -1001,6 +1018,7 @@ C
       COMMON /C0/ALP1,ALP2,ALP3,ALP4,ALP5,SD,CD,SDSD,CDCD,SDCD,S2D,C2D
       DATA F0,F1,F2,PI2/0.D0,1.D0,2.D0,6.283185307179586D0/
       DATA EPS/1.D-6/
+!$OMP THREADPRIVATE(/C0/)
 C-----
       ALP1=(F1-ALPHA)/F2
       ALP2= ALPHA/F2
@@ -1034,10 +1052,11 @@ C***** INPUT
 C*****   X,Y,D : STATION COORDINATES IN FAULT SYSTEM
 C### CAUTION ### IF X,Y,D ARE SUFFICIENTLY SMALL, THEY ARE SET TO ZERO
 C
-      COMMON /C0/DUMMY(5),SD,CD
+      COMMON /C0/DUMMY(5),SD,CD,DUMMY1(5)
       COMMON /C1/P,Q,S,T,XY,X2,Y2,D2,R,R2,R3,R5,QR,QRX,A3,A5,B3,C3,
      *           UY,VY,WY,UZ,VZ,WZ
       DATA  F0,F1,F3,F5,EPS/0.D0,1.D0,3.D0,5.D0,1.D-6/
+!$OMP THREADPRIVATE(/C0/, /C1/)
 C-----
       IF(DABS(X).LT.EPS) X=F0
       IF(DABS(Y).LT.EPS) Y=F0
@@ -1090,6 +1109,7 @@ C
       COMMON /C2/XI2,ET2,Q2,R,R2,R3,R5,Y,D,TT,ALX,ALE,X11,Y11,X32,Y32,
      *           EY,EZ,FY,FZ,GY,GZ,HY,HZ
       DATA  F0,F1,F2,EPS/0.D0,1.D0,2.D0,1.D-6/
+!$OMP THREADPRIVATE(/C2/)
 C-----
       IF(DABS(XI).LT.EPS) XI=F0
       IF(DABS(ET).LT.EPS) ET=F0
-- 
2.39.3

